internal/jobworker/job/job.go

```diff
@@ -0,0 +1,302 @@
+	"github.com/tjper/teleport/internal/jobworker/reexec"
+
+	"github.com/google/uuid"
+)
+
+// New creates a new Job instance.
+func New(
+	owner string,
+	cmd reexec.Command,
+) (*Job, error) {
+	var closers []io.Closer
+	cleanup := func() {
+		for _, closer := range closers {
+			closer.Close()
+		}
+	}
+
+	cmdOut, cmdIn, err := os.Pipe()
+	if err != nil {
+		return nil, fmt.Errorf("new job cmd pipe; error: %w", err)
+	}
+	closers = append(closers, cmdOut)
+	closers = append(closers, cmdIn)
+
+	continueOut, continueIn, err := os.Pipe()
+	if err != nil {
+		cleanup()
+		return nil, fmt.Errorf("new job continue pipe; error: %w", err)
+	}
+	closers = append(closers, continueOut)
+	closers = append(closers, continueIn)
+
+	shellCmd, err := os.Executable()
+	if err != nil {
+		cleanup()
+		return nil, fmt.Errorf("fetch current exec; error: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	executable := exec.CommandContext(ctx, shellCmd, jobworker.Reexec)
+	executable.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	executable.ExtraFiles = []*os.File{cmdOut, continueOut}
+
+	id := uuid.New()
+	logger.Infof("Constructed New Job; ID: %v", id)
+	return &Job{
+		mutex:       new(sync.RWMutex),
+		ID:          id,
+		Owner:       owner,
+		cmd:         cmd,
+		status:      Pending,
+		exitCode:    noExit,
+		ctx:         ctx,
+		cancel:      cancel,
+		exec:        executable,
+		cmdIn:       cmdIn,
+		cmdOut:      cmdOut,
+		continueIn:  continueIn,
+		continueOut: continueOut,
+	}, nil
+}
+
+// Job represents a single arbitrary command and its related entities
+// (output, status, etc.).
+type Job struct {
+	// TODO: Consider replacing general Job mutex with field specific mutexes to
+	// mitigate unnecessary lock contention.
+	mutex *sync.RWMutex
+
+	// ID is a unique identifier.
+	ID uuid.UUID
+	// Owner is the user responsible for Job instance creation.
+	Owner string
+
+	cmd      reexec.Command
+	status   Status
+	exitCode int
+
+	// context.Context is usually utilized at the function level. However, here
+	// it is being used to coordinate the cancelling of all async Job resources.
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	exec                    *exec.Cmd
+	cmdIn, cmdOut           io.WriteCloser
+	continueIn, continueOut io.WriteCloser
+}
+
+// StreamOutput streams Job's output to the passed stream channel in chunks of
+// size chunkSize. StreamOutput will return if either of the following
+// circumstances occur:
+//
+// 1) The ctx is cancelled.
+// 2) The Job is no longer running and the end of the output is reached.
+func (j Job) StreamOutput(ctx context.Context, stream chan<- []byte, chunkSize int) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	fd, err := os.Open(output.File(j.ID))
+	if err != nil {
+		return fmt.Errorf("open job output; error: %w", err)
+	}
+	go func() {
+		<-ctx.Done()
+		fd.Close()
+	}()
+
+	b := make([]byte, chunkSize)
+	for {
```

Comment:
A blocking StreamOutput function forces all callers to deal with concurrency for it to be useful. In this case I think making the function non-blocking would make for an API that is easier to use.

You may either want to return the channel for streaming or take ownership of it, so it can be closed when the stream ends. We'd get a usage idiom like this:


```go
ch := make(chan []byte)
if err := job.StreamOutput(ctx, ch, chunkSize); err != nil {
  // Abort.
}
for _, chunk := range ch {
  // Handle chunk
}
```

If you take the change don't forget to update the godocs.